### PR TITLE
asMap() not part of tensor interface anymore

### PR DESCRIPTION
**Pull request by kaifox**
* Introduced one new interfacce (MappableTensor)
* Additional static method mapFrom() in TensorInternals
* Delegation method to the above
* Only the immutable tensor still implements the MappableTensor interface, the others use the fallback solution (as they did already internally)

Please have a look if this looks ok to you, The branch is based on the last changes (context becomes position)

### CODE REVIEW

File: src/java/org/tensorics/core/tensor/MappableTensor.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * Enhances the tensor interface by a method to view the tensor as a map from
+ * position to values. This interface is intended to be used by utility methods
+ * in order to avoid unnecessary calls.
+ * 
+ * @author kaifox
+ *
+ * @param <V>
+ *            the type of the values of the tensors
+ */
+public interface MappableTensor<V> extends Tensor<V> {
```
Comment (michi42) [inline src/java/org/tensorics/core/tensor/MappableTensor.java]: What is the advantage of having MappableTensor extend Tensor, instead of Mappable being independent and ImmutableTensor implementing Tensor, Mappable?
Just to limit the number of levels in the class hierarchy, especially if we're going to introduce at some point other specialized interfaces like Sliceable, Sortable, etc. for optimization ...

File: src/test/org/tensorics/core/tensor/operations/TensorInternalsTest.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	@Test
+	public void mapIsConstructedFromShape() {
+		@SuppressWarnings("unchecked")
+		Tensor<String> tensor = mock(Tensor.class);
+		when(tensor.shape()).thenReturn(SHAPE_TO_RETURN);
+		when(tensor.get(Position.empty())).thenReturn(VALUE_TO_RETURN);
+
+		Map<Position, String> returned = TensorInternals.mapFrom(tensor);
+		assertThat(returned).isEqualTo(MAP_TO_RETURN);
+	}
```
Comment (michi42) [inline src/test/org/tensorics/core/tensor/operations/TensorInternalsTest.java]: perhaps consider adding a test with a non-zero-dimensional, non-mappable Tensor (as the 0-dim is a rather special case)

**Review by michi42 — COMMENTED** [review]
Very nice! The Mappable vs. MappableTensor we should decide on though ...

File: src/java/org/tensorics/core/tensor/MappableTensor.java (hunk 1 above)
Comment (kaifox) [reply src/java/org/tensorics/core/tensor/MappableTensor.java]: Agreed! I will change.

**Review by michi42 — APPROVED** [review]
Great, thanks!
